subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/DeprecatedUsageProgressDetails.java

```diff
@@ -27,11 +28,19 @@
  */
 @UsedByScanPlugin
 public interface DeprecatedUsageProgressDetails {
-    String getMessage();
 
-    String getDetails();
+    String getSummary();
 
+    String getRemovalDetails();
+
+    @Nullable
     String getAdvice();
 
+    @Nullable
+    String getContextualAdvice();
```

Comment:
What is the difference between "getAdvice" and "getContextualAdvice"?

Reply:
I will describe this in the code before we are done here.

The advice for a deprecation must be constant for the thing that was deprecated. The contextual advice can be dynamic based on the context. For example, the deprecation of loading annotation processors from the compile classpath has contextual advice indicating which processors were found.

This distinction allows us to much better group and visualise deprecations. We can group all instances of loading processors off the compile classpath, but give details about specific usages where appropriate

We have one other case of this where users extend FileCollection, where we indicate the name of the user class.

Reply:
I know we usually don't have much comment on the internal classes, but I think it's better to make these explanation into comments here. I was really confused by the name without looking at Luke's explanation.